src/material_library.cpp

```diff
@@ -0,0 +1,395 @@
+  s.resize(f.tellg());
+  f.seekg(0, std::ios::beg);
+  f.read(&s[0], s.size());
+  f.close();
+  Json::Reader reader;
+  Json::Value json;
+  reader.parse(s, json);
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f;
+  f.open(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat) {
+  // if exists, get the material name from metadata make one instead
+  std::string mat_name;
+  int mat_number = -1;
+  std::set<int>::iterator mat_numb_it;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue) {
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+    }
+  }
+  pyne::matname_set::iterator key_it;
+  if (mat.metadata.isMember("name")) {
+    if (Json::intValue <= mat.metadata["name"].type() &&
+        mat.metadata["name"].type() <= Json::realValue) {
+      mat_name = std::to_string(mat.metadata["name"].asInt());
+      mat.metadata["name"] = mat_name;
+    } else {
+      mat_name = mat.metadata["name"].asString();
+    }
+
+  } else {
+    if (mat_number == -1) {
+      mat_number = keylist.size();  // set a temp mat_number to form mat name
+    }
+    mat_name = "_" + std::to_string(mat_number);
+    mat.metadata["name"] = mat_name;
+  }
+
+  add_material(mat_name, mat);
+}
+
+void pyne::MaterialLibrary::add_material(const std::string& key,
+                                         pyne::Material mat) {
+  std::string mat_name;
+  int mat_number = -1;
+  std::set<int>::iterator mat_numb_it;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue) {
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+      mat.metadata["mat_number"] = mat_number;
+    }
+    mat_numb_it = mat_number_set.find(mat_number);
+    if (mat_numb_it != mat_number_set.end()) {
+      std::string msg = "Material number ";
+      msg += mat_number;
+      msg += " is already in the library.";
+      warning(msg);
+    }
+  }
+  if (!mat.metadata.isMember("name")) {
+    mat.metadata["name"] = key;
+  }
+
+  append_to_nuclist(mat);
+  if (mat_number > 0) mat_number_set.insert(mat_number);
+  std::pair<std::set<std::string>::iterator, bool> key_insert;
+  key_insert = keylist.insert(key);
+  // if the key was not present in the MaterialLibrary add it, otherwise
+  // overwrite the existing material
+  if (key_insert.second == true) name_order.push_back(key);
+  material_library[key] = new Material(mat);
```

Comment:
if the key exists is it not better to re-assign the value instead of deleting the pointer ?

Reply:
These are two separate things:
1. we should probably reassign the value
2. there is a risk of a memory leak because this could be the last reference to that material object??? not sure how to handle that.

Reply:
if we overwrite the value of the pointer I guess we are not missing the reference to the previous material object (the reference remain the same only the value changes ?)

Reply:
Before this line, `material_library[key]` may point to some `Material` object that was allocated on the heap; let's call that `pointer_to_old_mat_with_name_key`.

After this line, `material_library[key]` will point to some new `Material` object, let's call that `pointer_to_new_mat_with_name_key`.

There may be no other variable in the code that knows about `pointer_to_old_mat_with_name_key` and this this memory becomes lost/leaked.  I think the only solution is to use reference counting points, but we haven't introduced anything like that in PyNE before.

Reply:
I might have missed your point:
 I understand that with the correct implementation we have a memory leak if we replace a `Material_ptr` that was already presents.

But I am still wondering why we can't just do something like:
```
if (exists)
    *material_library[key] = new_mat
else:
    material_library[key] = new_mat_ptr
```
(the issue with this, would be that we would change the value of all the material_ptr that point to the same address as the `material_library[key]` pointer)

using `shared_ptr` would definitely solve that for us.
I thing it is a good time to introduce `shared_ptr`, particularly because this implementation is new, so we won't have to bother introducing them later...
we already require C++11, let's use all of its capabilities....

Reply:
@gonuke I witched to shared_ptr, I actually think this is best for shared python/c++ library and will avoid future problem about who owns the pointer (the python or the C++).

I think it would be a great improvement in robustness to make the change everywhere.... (a great Hack-a-thon topic?)